### PR TITLE
DPO trainer supports num_logits_to_keep to save memory

### PR DESCRIPTION
**Pull request by xyangk**
# What does this PR do?

Since transformers v4.45.0 supports `num_logits_to_keep` in the forward pass(check this [PR](https://github.com/huggingface/transformers/pull/31292)), it can be useful for saving memory and speeding up training by computing logits only for the last completion tokens, especially in scenarios with very long prompts where labels are ignored (-100).

I implemented the logic to calculate `num_logits_to_keep` based on the labels and tested this on DPO, resulting in substantial memory savings.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by xyangk** [comment]
Sure!

**Comment by xyangk** [comment]
Implemented the calculation of num_logits_to_keep on the new master code, fixed a bug in the previous calculation, and used more rigorous test code.

**Comment by xyangk** [comment]
Hi @kashif , I wanted to check in on this PR. Please let me know if there's anything further I can clarify or update. Thank you for your time and review!

**Comment by kashif** [comment]
@xyangk can you use the transformers helper `check_min_version` instead of implementing the logic yourself?

**Comment by xyangk** [comment]
> @xyangk can you use the transformers helper `check_min_version` instead of implementing the logic yourself?

I've implemented the suggested changes. Thank you for the feedback!

**Comment by kashif** [comment]
@xyangk  in view of the fix https://github.com/huggingface/trl/pull/2332 can you kindly confirm that this feature is not affected by this?

**Comment by xyangk** [comment]
> in view of the fix #2332 can you kindly confirm that this feature is not affected by this?

I've reviewed the code in realtion to fix #2332. I can confirm that this feature remains unaffected by those changes. I've also added some comments to improve clarity.

**Review by kashif — APPROVED** [review]

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -529,6 +530,11 @@ def make_inputs_require_grad(module, input, output):
                 )
 
             self.add_callback(SyncRefModelCallback(ref_model=self.ref_model, accelerator=self.accelerator))
+
+        # num_logits_to_keep is supported since transformers v4.45.0
+        if self.use_num_logits_to_keep:
+            check_min_version("4.45.0")
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:536]: You don't need it

https://github.com/huggingface/trl/blob/main/setup.py#L82

Comment (xyangk) [reply trl/trainer/dpo_trainer.py:536]: Sorry I missed that the transformers version was already updated. Since the current PR is already merged, I'll create a new PR to remove the version check. Thank you for cathing this!

Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:536]: #2343
